pkg/leader/leader.go

```diff
@@ -0,0 +1,191 @@
+)
+
+// ErrNoNS indicates that a namespace could not be found for the current
+// environment
+var ErrNoNS = errors.New("namespace not found for current environment")
+
+// TryBecome behaves like Become, except it will not return an error in the
+// case where a namespace cannot be found for the current pod. This is useful
+// for a service that might run outside the cluster, for example an operator
+// being started with `operator-sdk up local`.
+func TryBecome(ctx context.Context, name string) error {
+	err := Become(ctx, name)
+	if err == ErrNoNS {
+		logrus.Warn("leader election disabled; no namespace was detected")
+		return nil
+	}
+	return err
+}
+
+// Become ensures that the current pod is the leader within its namespace. It
+// continuously tries to create a ConfigMap with the provided name and the
+// current pod set as the owner reference. Only one can exist at a time with
+// the same name, so the pod that successfully creates the ConfigMap is the
+// leader. Upon termination of that pod, the garbage collector will delete the
+// ConfigMap, enabling a different pod to become the leader.
+func Become(ctx context.Context, name string) error {
+	logrus.Info("trying to become the leader")
+
+	ns, err := myNS()
+	if err != nil {
+		return err
+	}
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return err
+	}
+
+	client, err := crclient.New(config, crclient.Options{})
+	if err != nil {
+		return err
+	}
+
+	owner, err := myOwnerRef(ctx, client, ns)
+	if err != nil {
+		return err
+	}
+
+	cm := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            name,
+			Namespace:       ns,
+			OwnerReferences: []metav1.OwnerReference{owner},
+		},
+	}
+
+	// check for existing lock from this pod, in case we got restarted
+	existing := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+	}
+	key := crclient.ObjectKey{Namespace: ns, Name: name}
+	err = client.Get(ctx, key, existing)
+
+	switch {
+	case err == nil:
+		for _, existingOwner := range existing.GetOwnerReferences() {
+			if existingOwner.Name == owner.Name {
+				logrus.Info("Found existing lock with my name. I was likely restarted.")
+				logrus.Info("Continuing as the leader.")
+				return nil
+			} else {
+				logrus.Infof("Found existing lock from %s", existingOwner.Name)
+			}
+		}
+	case apierrors.IsNotFound(err):
+		logrus.Info("No pre-existing lock was found.")
+	default:
+		logrus.Error("unknown error trying to get ConfigMap")
+		return err
+	}
+
+	// try to create a lock
+	seconds := 1
+	for {
+		err := client.Create(ctx, cm)
+		switch {
+		case err == nil:
+			logrus.Info("Became the leader.")
+			return nil
+		case apierrors.IsAlreadyExists(err):
+			logrus.Info("Not the leader. Waiting.")
+			select {
+			case <-time.After(wait.Jitter(time.Second*time.Duration(seconds), .2)):
+				if seconds < 16 {
```

Comment:
maybe change `16` to a named variable e.g `maxBackoffInterval` and also use `backoff`
```go
if backoff < maxBackoffInterval {
   backoff *= 2
}
```